flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/TtlStateTestBase.java

```diff
@@ -471,6 +473,26 @@ public void testRestoreTtlAndRegisterNonTtlStateCompatFailure() throws Exception
         sbetc.createState(ctx().createStateDescriptor(), "");
     }
 
+    @Test
+    public void testIncrementalCleanupWholeState() throws Exception {
+        assumeTrue(incrementalCleanupSupported());
+        initTest(getConfBuilder(TTL).cleanupIncrementally(5, true).build());
+        timeProvider.time = 0;
+        // create enough keys to trigger incremental rehash
+        updateKeys(0, INC_CLEANUP_ALL_KEYS, ctx().updateEmpty);
+        // expire all state
+        timeProvider.time = 120;
+        // trigger state clean up
+        for (int i = 0; i < INC_CLEANUP_ALL_KEYS; i++) {
+            sbetc.setCurrentKey(Integer.toString(i));
+        }
+        // check all state cleaned up
+        for (int i = 0; i < INC_CLEANUP_ALL_KEYS; i++) {
+            sbetc.setCurrentKey(Integer.toString(i));
+            assertTrue("Original state should be cleared", ctx().isOriginalNull());
```

Comment:
I think the kernel logic is to verify that current `StateTable` at `currentNamespace` is null and current logic seems a bit strange as its name `is original null` hided what we want.

I prefer to change this line to
~~~java
assertTrue("Original state table should be cleared", isCurrentStateTableNull());
~~~

And we could also adding method below:
~~~java
    @SuppressWarnings({"rawtypes", "unchecked"})
    private boolean isCurrentStateTableNull() {
        InternalKvState<?, String, ?> original = ((AbstractTtlDecorator<InternalKvState<?, String, ?>>) ctx.ttlState).original;
        Preconditions.checkState(original instanceof AbstractHeapState);
        return ((AbstractHeapState) original).getStateTable().get(ctx.currentNamespace) == null;
    }
~~~

By doing so, we could remove those strange `isValueOfCurrentKeyNull` and `isOriginalNull` methods.

Reply:
I agree it is somewhat over-designed here but the original thought is to try not to fail the test if `assumeTrue(incrementalCleanupSupported())` returns `true` for RocksDB backend in the future.

The suggested way works but the name `isCurrentStateTableNull` is misleading... It's "the state of the mapping for the composite of active key and given namespace" (referring to `StateTable#get(N namespace)`) that should be null, rather than the current state table. `isOriginalCleared` may be a better name.

Reply:
Here we're testing the existence of the state of specific key and namespace, right? so why not rename the method to `isOriginalExisting()`, which can also keep the same style with another method `isOriginalEmptyValue()`.

Reply:
My original point is that we could avoid to introduce those `isValueOfCurrentKeyNull` related methods by just introduce one single `isCurrentStateTableNull` method within `TtlStateTestBase.java` (no matter what name it is). 

FLINK-21413 is about to fix a performance problem to avoid many unnecessary empty hash map and state entry, instead of incorrect query result. This new added test should only targets for heap-based keyed state backend and we can safely cast to `AbstractHeapState` here.

Reply:
@Myasuka  Oh.. I see what you mean here. Very good point. :) 

I've improved the codes based on your comments.